lib/bolt/result_set.rb

```diff
@@ -0,0 +1,115 @@
+module Bolt; class ResultSet
+  attr_reader :results
+
+  include Enumerable
+
+  # We only want want to include these when puppet is loaded
+  def self.include_iterable
+    include(Puppet::Pops::Types::Iterable)
+    include(Puppet::Pops::Types::IteratorProducer)
+  end
+
+  def iterator
+    # TODO: does this work outside a compiler?
+    if Object.const_defined?(:Puppet) && Puppet.const_defined?(:Pops) &&
+        self.class.included_modules.include?(Puppet::Pops::Types::Iterable)
+      Puppet::Pops::Types::Iterable.on(@results, Bolt::Result)
+    else
+      @results.each
+    end
+  end
+
+  def initialize(results)
+    @results = results
+  end
+
+  def each
+    @results.each { |r| yield r }
+  end
+
+  def result_hash
+    @result_hash ||= @results.reduce({}) do |acc, result|
+      # TODO: What is the ID of a result target
+      acc[result.target.uri] = result
+      acc
+    end
+  end
+
+  def count
+    @results.size
+  end
+  alias length count
+  alias size count
+
+  def empty
+    @results.empty?
+  end
+  alias empty? empty
+
+  # Should this be nodes or targets?
+  def targets
+    results.map(&:target)
+  end
+
+  # Is this useful? For logging maybe?
+  def names
+    @results.map { |r| r.target.name }
+  end
+
+  def ok
+    @results.all? { |r| r.ok? }
+  end
+  alias ok? ok
+
+  def errorset
+    filtered = @results.select { |r| !r.ok? }
+    ResultSet.new(filtered)
+  end
+
+  def okset
+    filtered = @results.select { |r| r.success? }
+    self.class.new(filtered)
+  end
+
+  def find(target_name)
+    result_hash[target_name]
+  end
+
+  def first
+    @results.first
+  end
+
+  # TODO: should this exist?
+  def [](node_uri)
```

Comment:
Asking whether you should be able to look a value up by its name/uri? I feel like the one case that could be useful is when cross-referencing between several results.  Like: several nodes performed an action against load-balanced servers, I want to group the results of that action by the server they operated against.

I feel like I'm having to stretch for examples, but it doesn't feel like much work to enable some use-cases we can't envision right now.

Reply:
I guess `find` already allows that. Are you suggesting that as better than `[]`?